Dockerfile

```diff
@@ -45,23 +45,37 @@ WORKDIR ${PROJECT_DIR}
 # Copy local source over
 COPY . ${PROJECT_DIR}
 
-# Get google dp dependency
-RUN mkdir -p third_party && \
-    cd third_party && \
-    git clone https://github.com/google/differential-privacy.git
+# Dealing with 3rd party code one needs to pin it to a version or a SHA.
+# Either we clone it here as was done but then it should be @e819e03 to match the submodule 
+# and will needed to be updated in keeping with that (a rather tedious way)
+# Get google dp dependency. Rather copy the thrid party deps as it's a cloned submodule
+# COPY ./third_party third_party
 
+# RUN mkdir -p third_party && \
+#     cd third_party && \
+#     git clone https://github.com/google/differential-privacy.git
+
+# Or we just copy it in and ensure the CI/CD clone the repo and the submodule.
```

Comment:
Please see the build it all passes as I pull all submodules 
In github actions one just need to do:
```
steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
```
"So what we can do here is probably stick to mentioning the sha." 
That was the alternative suggested but it's very hidden I believe it'll break as people **will** forget. 

FYI:
If you want to clone a repo with it's submodules (Which is what I do on my local build) simply use:
```
git clone --recurse-submodules https://github.com/OpenMined/PyDP.git
``` 
This will correctly clone the repo at the sha of upstream as shown in the repo.

Reply:
That is not a problem as we clone the submodules in the github action already 
See the github action file.
```
steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
```

As for the local builds one just needs to clone the submodules using `--recurse-submodules `

FYI:
For running things locally or in any other checkouts one just needs to clone as follows. 
```
git clone --recurse-submodules https://github.com/OpenMined/PyDP.git
```
This will correctly clone the upstream dependencies (`third_party`) at the correct sha. 

"So what we can do here is probably stick to mentioning the sha."
This is an alternative as suggested however, it is a very hidden config an **will** be forgotten and things will break (or not break when they should) as that sha will become out of sync with the one mentioned in the submodule. I'm in favor of letting git handle as much as possible.

However the remaining issues do still remain.
https://github.com/OpenMined/PyDP/pull/176#issuecomment-641100600

Reply:
Oh, I didn't notice the submodule in actions.  
For running locally, we already have scripts that take cares of this stuff.

Reply:
> 1. Are the image names and tags correct? Currently  openmined/pydp-{branch}-{python-version} Is this consistent across the organization?

This looks good. 

> 1. Un-comment the last stage to push the images to docker hub and add the credentials to github secrets.
> 2. Remove verbose comments and scratch marks in Dockerfile. I left it to assist the reviewer.

I'll make these changes while adding the credentials